fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/HNSW.java

```diff
@@ -0,0 +1,2022 @@
+                                nodeReference.getPrimaryKey()))
+                .thenApply(node -> biMapFunction.apply(nodeReference, node));
+    }
+
+    /**
+     * Asynchronously fetches neighborhood nodes and returns them as {@link NodeReferenceWithVector} instances,
+     * which include the node's vector.
+     * <p>
+     * This method efficiently retrieves node data by first checking an in-memory {@code nodeCache}. If a node is not
+     * in the cache, it is fetched from the {@link StorageAdapter}. Fetched nodes are then added to the cache to
+     * optimize subsequent lookups. It also handles cases where the input {@code neighborReferences} may already
+     * contain {@link NodeReferenceWithVector} instances, avoiding redundant work.
+     *
+     * @param <N> the type of the node reference, extending {@link NodeReference}
+     * @param storageAdapter the storage adapter to fetch nodes from if they are not in the cache
+     * @param readTransaction the transaction context for database read operations
+     * @param storageTransform an affine transformation operator that is used to transform the fetched vector into the
+     *        storage space that is currently being used
+     * @param layer the graph layer from which to fetch the nodes
+     * @param neighborReferences an iterable of references to the neighbor nodes to be fetched
+     * @param nodeCache a map serving as an in-memory cache for nodes. This map will be populated with any
+     * nodes fetched from storage.
+     *
+     * @return a {@link CompletableFuture} that, upon completion, will contain a list of
+     * {@link NodeReferenceWithVector} objects for the specified neighbors
+     */
+    @Nonnull
+    private <N extends NodeReference> CompletableFuture<List<NodeReferenceWithVector>>
+            fetchNeighborhood(@Nonnull final StorageAdapter<N> storageAdapter,
+                              @Nonnull final ReadTransaction readTransaction,
+                              @Nonnull final AffineOperator storageTransform,
+                              final int layer,
+                              @Nonnull final Iterable<? extends NodeReference> neighborReferences,
+                              @Nonnull final Map<Tuple, Node<N>> nodeCache) {
+        return fetchSomeNodesAndApply(storageAdapter, readTransaction, storageTransform, layer, neighborReferences,
+                neighborReference -> {
+                    if (neighborReference instanceof NodeReferenceWithVector) {
+                        return (NodeReferenceWithVector)neighborReference;
+                    }
+                    final Node<N> neighborNode = nodeCache.get(neighborReference.getPrimaryKey());
+                    if (neighborNode == null) {
+                        return null;
+                    }
+                    return new NodeReferenceWithVector(neighborReference.getPrimaryKey(),
+                            neighborNode.asCompactNode().getVector());
```

Comment:
I just want to be explicit about this: is the supposition that it's safe to call `.asCompactNode()` here coming from the fact that if the node was an `InliningNode`, then we wouldn't reach here (or the line below in the case where we actually have to look up the node)? It might be good to comment that

Reply:
done.